recipe/meta.yaml

```
@@ -1,48 +1,73 @@
 {% set name = "curl-cffi" %}
-{% set version = "0.11.4" %}
+{% set version = "0.12.1b2" %}
 
 package:
   name: {{ name|lower }}
   version: {{ version }}
 
 source:
-  url: https://pypi.org/packages/source/{{ name[0] }}/{{ name }}/curl_cffi-{{ version }}.tar.gz
-  sha256: 064ecab44c0571aaf47ad18a97200130f9e05bf7668fe3a2ff8633276fcbbb3c
+  url: https://pypi.org/packages/source/{{ name[0] }}/{{ name }}/{{ name|lower|replace('-', '_') }}-{{ version }}.tar.gz
+  sha256: 9a2d0c9e164c0425e6bd610455e188a1b8e952392f90cef02e794589a7f02c8f
```

You might recall I said that it downloads `curl-impersonate`:

```
Downloading libcurl-impersonate from https://github.com/lexiforest/curl-impersonate/releases/download/v1.1.2/libcurl-impersonate-v1.1.2.x86_64-linux-gnu.tar.gz
```

where we probably want to download that ourselves and patch the build script to use our downloaded version

```suggestion
  - url: https://pypi.org/packages/source/{{ name[0] }}/{{ name }}/{{ name|lower|replace('-', '_') }}-{{ version }}.tar.gz
    sha256: 9a2d0c9e164c0425e6bd610455e188a1b8e952392f90cef02e794589a7f02c8f
  - url: https://github.com/lexiforest/curl-impersonate/releases/download/v1.1.2/libcurl-impersonate-v1.1.2.x86_64-linux-gnu.tar.gz
  sha256: abc123
  patches:
    - 0001-use-downloaded-curl-impersonate.patch
```

---

@ifitchet It wasn’t easy, but I’ve got a draft ready. The most challenging part is that macOS binaries have hardcoded paths to dynamic libraries. I tried to work around this using `install_name_tool`—though I’m not sure if that’s the correct approach. You can take a look at my meta.yaml and let me know what you think. Ignore the comments — I just included everything to illustrate the point.

The macOS `libcurl-impersonate.4.dylib` for `x86_64` has one dependency that’s not present in our recipes—and even conda-forge doesn’t have it—which is `librtmp.1.dylib`. You can see the latest failed logs from Prefect. I can't just add it to `missing_dso_whitelist` because the tests would fail anyway.

I’m wondering if it’s reasonable to drop this package for `[osx and x86_64]`.

---

@cbouss Could you share your opinion on this? libcurl-impersonate is a library that the upstream downloads manually. We're trying to download it using conda-build and then reuse it (see added patch). Please read this conversation from the beginning for more details.

---

Hmm, I just went down a rabbit hole of _where are the shared libraries anyway?_

It turns out that in macOS `/usr/lib/X.dylib` are now mapped into large dynamic objects, you can

```
$grep ncurses /System/Volumes/Preboot/Cryptexes/OS/System/Library/dyld/*.map
/System/Volumes/Preboot/Cryptexes/OS/System/Library/dyld/dyld_shared_cache_arm64e.map:/usr/lib/libncurses.5.4.dylib
/System/Volumes/Preboot/Cryptexes/OS/System/Library/dyld/dyld_shared_cache_x86_64.map:/usr/lib/libncurses.5.4.dylib
```
and looking closer those maps have pointers into the adjacent dynamic objects.

That doesn't answer your question about `librtmp` as that doesn't appear there.

---

The `install_name_tool` rewrite is the right approach for hard-coded pathnames -- it used elsewhere in recipes (always comforting to know you're not breaking new ground!)

On a linux-aarch64 build, the checking is only run against `lib/python3.10/site-packages/curl_cffi/_wrapper.abi3.so` (which appears to contain symbols for boringssl amongst other which is part of `curl-impersonate`) whereas on `osx-arm64` the checking is done against both `lib/python3.11/site-packages/curl_cffi/_wrapper.abi3.so` and `lib/python3.11/site-packages/extralib/libcurl-impersonate.4.dylib` (which may be an artifact of macOS in that it can't/chooses not to slurp objects from one library to another)

But it's not showing `librtmp`.  Oh wait, that's a fail on `osx-64` only.  Maybe that `curl-impersonate` tarball is built differently?

---

So they are:

```
$otool -L libcurl-impersonate-{64,arm64}/libcurl-impersonate.dylib 
libcurl-impersonate-64/libcurl-impersonate.dylib:
	/Users/runner/work/_temp/install/lib/libcurl-impersonate.4.dylib (compatibility version 13.0.0, current version 13.0.0)
	/usr/local/opt/libidn2/lib/libidn2.0.dylib (compatibility version 5.0.0, current version 5.0.0)
	/usr/local/opt/rtmpdump/lib/librtmp.1.dylib (compatibility version 0.0.0, current version 0.0.0)
	/System/Library/Frameworks/LDAP.framework/Versions/A/LDAP (compatibility version 1.0.0, current version 2.4.0)
	/usr/local/opt/zstd/lib/libzstd.1.dylib (compatibility version 1.0.0, current version 1.5.7)
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.12)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 1600.157.0)
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 2202.0.0)
	/System/Library/Frameworks/CoreServices.framework/Versions/A/CoreServices (compatibility version 1.0.0, current version 1226.0.0)
	/System/Library/Frameworks/SystemConfiguration.framework/Versions/A/SystemConfiguration (compatibility version 1.0.0, current version 1296.60.3)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1336.61.1)
libcurl-impersonate-arm64/libcurl-impersonate.dylib:
	/Users/runner/work/_temp/install/lib/libcurl-impersonate.4.dylib (compatibility version 13.0.0, current version 13.0.0)
	/opt/homebrew/opt/libidn2/lib/libidn2.0.dylib (compatibility version 5.0.0, current version 5.0.0)
	/System/Library/Frameworks/LDAP.framework/Versions/A/LDAP (compatibility version 1.0.0, current version 2.4.0)
	/opt/homebrew/opt/zstd/lib/libzstd.1.dylib (compatibility version 1.0.0, current version 1.5.7)
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.12)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 1700.255.5)
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 2503.1.0)
	/System/Library/Frameworks/CoreServices.framework/Versions/A/CoreServices (compatibility version 1.0.0, current version 1226.0.0)
	/System/Library/Frameworks/SystemConfiguration.framework/Versions/A/SystemConfiguration (compatibility version 1.0.0, current version 1300.120.2)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1345.120.2)
```

How, inconvenient.

Also, looking on a personal osx-64 box, `librtmp` doesn't appear in those maps.

---

I'm also curious why `libcurl-impersonate.dylib` requires `librtmp`. It doesn't appear to be ARM-specific, and the osx-64 version doesn't include it.

---

But yeah, this is macOS — nothing surprises me anymore after these past two days. 😅

---

It could be that `curl-impersonate` was built on an older/newer version of macOS Intel which pulls/pulled `librtmp` into the mix.
This is where vendoring is bad and vendoring binaries is doubly bad.

Before you get excited, as noted elsewhere, I did try to build `curl-impersonate` itself although it failed to find things that it was vendoring.
Although, if this proves intractable, then maybe that's our option.  🤷

---

So, what should I do next? Maybe I missed your point, but you mentioned you tried building `curl-impersonate` and it failed. I have a Mac with an ARM chip, so I’ll probably need to create a virtual environment to try building the osx-64 version of `curl-impersonate`.

I'm wondering — is it okay to start a conversation with the upstream maintainers at this point? Either way, even if they do provide a newer version of `curl-impersonate` without `librtmp` (if it’s really related to an old Intel-based macOS), we still need version 1.0.0 to match `curl-cffi 0.12.0` requirement. So... yeah, probably not the way to go.

So, do you think about creation of `curl-impersonate-feedstock`? Personally, I see curl-impersonate as a bit of a vendor mess. Looks like we're going down a rabbit hole of unvendoring stuff.

---

When I looked at `curl-impersonate` (I started a local feedstock using grayskull maybe) it was vendoring about four things but _we_ didn't have packages for those so the vendoring itself wasn't so much of an issue.

The issue was that it would download and build `boringssl`, say, do something else, then immediately say "I can't find boringssl" -- huh?  You've just built it!

I didn't have time to investigate.

Given that the `osx-64` `curl-impersonate` binary looks like it will cause us problems then maybe we should look at building `curl-impersonate` ourselves.
Maybe the problem with `boringssl` (or whatever) will be easily fixed.